### PR TITLE
feat(#72): implement GPX and TCX encoding

### LINKED ISSUE #72: Implement GPX and TCX encoding
Currently, when we edit the activity using Tools that we have, we save them as FIT file format.  We need to add more flexibility to the users to select the file format based on their need.

Requirements:
1. (WASM) Implement the GPX and TCX encoding
1. (UI) in `Tools`: add dropdown menu to select file format: FIT, GPX, TCX right below Tool selection dropdown.
    - add help text something like this if user selects GPX or TCX "If your target platform is Strava, please choose FIT instead. With GPX or TCX, the target device name may not map correctly with Strava's device mapping database, as they have shifted their support to FIT files."
1. (WASM-UI): for GPX and TCX, target device will be plain input text mapped to `deviceName` in `EncodeSpecifications` class/struct and is ~~not~~ a mandatory.
1. (UI) Since GPX and TCX have their own limitations, display help text indicating which fields will be omitted (~~only if those fields exist in the currently opened activities~~).
    - GPX doesn't support power (ref: [schema](https://www.topografix.com/gpx.asp), [extension](https://www8.garmin.com/xmlschemas/TrackPointExtensionv1.xsd))
    - TCX doesn't support temperature  (ref: [schema](https://www8.garmin.com/xmlschemas/TrainingCenterDatabasev2.xsd), [extension](https://www8.garmin.com/xmlschemas/ActivityExtensionv2.xsd))

### PR DESCRIPTION
**Pull request by muktihari**
- Fix preprocessing bug on GPX and TCX, the distance was not calculated correctly. Preprocessing should be call for all records per session instead of per lap.
   After we have activity with multiple lap/session converted to GPX and TCX files, the bug come into surface, now it has been fixed.
- Implement GPX and TCX encoding
- Implement UI component for selecting file type
- Tested: The exported file with multiple lap is successfully uploaded to Strava.

Overview:
image

image

image

image

Now the button change based on the selected file type:
image

This should close #72 after merged.

### CODE REVIEW

**Comment by muktihari** [comment]
Using one space indent reduce file size significantly from 19 MB to 16 MB while keep maintaining the readability.

image

We can further reduce the size if we minify the file (will be 13MB), this probably can be a new feature later but why we should not do it by default:
- If we minify by default, the file will be unreadable. A question will be asked "why not choose FIT instead if we want the most compact size?". 
 
So I think one space indent is the most suitable choice.

**Review by raditzlawliet — COMMENTED** [review]
I have tried but limited: 
- Pick .fit files
- Edit Relevant
- No change on other options (including target device)
- Convert to Tcx, Gpx and vice versa. 

All looks good, all files can opened back and convert agian on another 

But sometime weird on target device name inputfield. My original  .fit using `XOSS G+ (11)`
- From .fit to .fit Detected as XOSS G+ (11),
- From .fit to .tcx -> detect as openivity web (i think it's limitation of tcx)
- From .fit to .gpx Detected as XOSS G+ (11).

1. The missing part is when i try to open .gpx (from .fit), 
- **after i choose Target .FIT, the target device is not automatically selected.**
- after i choose Target .GPX, the target device automatically selected XOSS G+ (11)
image

2. When i try to convert .fit to .tcx. Target Device is automatically filling XOSS G+ (11), but when i try to reopened the .tcx, its `openitivy.github.io`. Is .tcx support target device? If .tcx doesn't support target device, why dont we hide the input?

File: src/components/ToolDeviceNameInput.vue [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  data() {
+    return {
+      deviceName: ''
+    }
+  },
+  computed: {
+    deviceNameFromFile(): string {
+      let deviceName = ''
+      for (let i = 0; i < this.activities.length; i++) {
+        if (deviceName != '') break
```
[the remark this answers is not in the document]
Comment (muktihari) [reply src/components/ToolDeviceNameInput.vue:35]: no, sure, I can change that

File: src/wasm/activity-service/activity/gpx/schema/metadata.go [hunk 2]
```diff
@@ -59,12 +62,58 @@ func (m *Metadata) UnmarshalXML(dec *xml.Decoder, se xml.StartElement) error {
...
+	if err := m.Link.Validate(); err != nil {
+		return fmt.Errorf("link: %w", err)
+	}
 	if err := m.Author.Validate(); err != nil {
 		return fmt.Errorf("author: %w", err)
 	}
 	return nil
 }
 
+var _ xml.Marshaler = &Metadata{}
```
[the remark this answers is not in the document]
Comment (muktihari) [reply src/wasm/activity-service/activity/gpx/schema/metadata.go:75]: This is like "Metadata implements xml.Marshaler"

This is to ensure that when I write the MarshalXML method, it should satisfy xml.Marshaler interface, if I did a typo then the vscode will complain right away. This is a common practice in writing go code.

**Comment by muktihari** [comment]
> after i choose Target .FIT, the target device is not automatically selected.
Yes, because **.fi**t uses fields: Manufacturer (uint16) and Product (uint16) for the device name, and **.gpx** uses plain text Creator (string), so there is no way to map back from a single string into two uint16 (Manufacturer and Product), even if we do tokenisation, I think it will bring complexity without added value.

**Comment by muktihari** [comment]
> From .fit to .tcx -> detect as openivity web (i think it's limitation of tcx)

for **.tcx** we have

```
type Device struct {
	Name      string   `xml:"Name"`
	UnitId    uint32   `xml:"UnitId"`
	ProductID uint16   `xml:"ProductId"`
	Version   *Version `xml:"Version"`
}
```

But this field is related only to Garmin Training Center App, I don't have full context of this and I think it doesn't map to fit directly. Strava drop support for gpx and tcx device mapping, it now use fit. And for those who want to add their device mapped correctly it enforce the manufacturer to upload it to its platform. ref: [https://developers.strava.com/docs/uploads/](https://developers.strava.com/docs/uploads/)

**Comment by muktihari** [comment]
> When i try to convert .fit to .tcx. Target Device is automatically filling XOSS G+ (11), but when i try to reopened the .tcx, its openitivy.github.io. Is .tcx support target device? If .tcx doesn't support target device, why dont we hide the input?

You found a bug in decoding tcx, it should have take the deviceName from Creator, but it took the deviceName for the Author. I will update this in different PR.

**Comment by raditzlawliet** [comment]
> > after i choose Target .FIT, the target device is not automatically selected.
> 
> Yes, because **.fi**t uses fields: Manufacturer (uint16) and Product (uint16) for the device name, and **.gpx** uses plain text Creator (string), so there is no way to map back from a single string into two uint16 (Manufacturer and Product), even if we do tokenisation, I think it will bring complexity without added value.
> 
> In fit, the device name is saved in FileId Mesg. ref: https://github.com/muktihari/fit/blob/master/profile/mesgdef/file_id_gen.go

I think doing compare string is enough (if not found just do like now) 

Since its clear, lgtm

**Review by raditzlawliet — APPROVED** [review]

**Comment by muktihari** [comment]
> I think doing compare string is enough (if not found just do like now)

Oh, I think I know what you mean, let me wrap thing up related that device name issue and will update in separate PR.
Will merge this first.

**Pull request merged**